### PR TITLE
[Api][Cart] Cart remove items

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

Screenshot 2020-06-30 at 23 10 31

### CODE REVIEW

File: features/cart/shopping_cart/removing_cart_item.feature [hunk 1]
```diff
@@ -7,25 +7,35 @@ Feature: Removing cart item from cart
...
+        And the store has a product "T-shirt strawberry" priced at "$17.22"
         And I add this product to the cart
         When I remove product "T-shirt banana" from the cart
         Then my cart's total should be "$17.22"
+
+    @api
+    Scenario: Checking cart's total after removing one item
+        Given the store has a product "T-shirt strawberry" priced at "$17.22"
+        And I added product "T-shirt banana" to the cart
+        And I add "T-shirt strawberry" product to the cart
```
Comment (GSadee) [inline features/cart/shopping_cart/removing_cart_item.feature:39]: Be consistent with naming the steps

File: features/cart/shopping_cart/removing_cart_item.feature [hunk 2]
```diff
@@ -7,25 +7,35 @@ Feature: Removing cart item from cart
     Background:
         Given the store operates on a single channel in "United States"
         And the store has a product "T-shirt banana" priced at "$12.54"
-        And I added product "T-shirt banana" to the cart
 
-    @ui
+    @ui @api
     Scenario: Removing cart item
+        Given I added product "T-shirt banana" to the cart
```
Comment (GSadee) [inline features/cart/shopping_cart/removing_cart_item.feature:13]: Why did you move this step from background as it is use in every scenario?

File: features/cart/shopping_cart/removing_cart_item.feature [hunk 3]
```diff
@@ -7,25 +7,35 @@ Feature: Removing cart item from cart
...
     Scenario: Checking cart's total after removing one item
-        Given the store has a product "T-shirt strawberry" priced at "$17.22"
+        Given I added product "T-shirt banana" to the cart
+        And the store has a product "T-shirt strawberry" priced at "$17.22"
         And I add this product to the cart
         When I remove product "T-shirt banana" from the cart
         Then my cart's total should be "$17.22"
+
+    @api
+    Scenario: Checking cart's total after removing one item
```
Comment (GSadee) [inline features/cart/shopping_cart/removing_cart_item.feature:36]: Why did you add a separate scenario for @api for this case?

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 4]
```diff
@@ -90,6 +100,18 @@ public function myCartShouldBeCleared(): void
...
+    /**
+     * @Then /^my (cart)'s total should be ("[^"]+")$/
+     */
+    public function myCartSTotalShouldBe(string $tokenValue, int $total): void
+    {
+        $response = $this->cartsClient->show($tokenValue);
+
+        $responseTotal = $this->responseChecker->getValue($response, 'total');
+
+        Assert::same($total, (int) $responseTotal);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:112]: Could be in one line

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 5]
```diff
@@ -185,6 +207,18 @@ private function putProductToCart(ProductInterface $product, string $tokenValue,
         $this->cartsClient->executeCustomRequest($request);
     }
 
+    private function removeProductFromCart(ProductInterface $product, string $tokenValue, int $quantity = 1): void
+    {
+        $request = Request::customItemAction('orders', $tokenValue, HttpRequest::METHOD_PATCH, 'remove');
+
+        $request->updateContent([
+            'productCode' => $product->getCode(),
+            'quantity' => $quantity,
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:216]: Why did you pass here a quantity?

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Cart/RemoveItemFromCartHandler.php [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        ProductVariantResolverInterface $variantResolver
+    ) {
+        $this->orderRepository = $orderRepository;
+        $this->productRepository = $productRepository;
+        $this->orderModifier = $orderModifier;
+        $this->orderProcessor = $orderProcessor;
+        $this->variantResolver = $variantResolver;
+    }
+
+    public function __invoke(RemoveItemFromCart $addItemToCart): OrderInterface
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Cart/RemoveItemFromCartHandler.php:58]:
```suggestion
    public function __invoke(RemoveItemFromCart $removeItemFromCart): OrderInterface
```

File: features/cart/shopping_cart/removing_cart_item.feature (hunk 2 above)
Comment (Tomanhez) [reply features/cart/shopping_cart/removing_cart_item.feature:13]: Because if I use step from api context between setup context steeps I have problem with doctrine like this: 
Screenshot 2020-06-29 at 11 41 07

File: features/cart/shopping_cart/removing_cart_item.feature (hunk 3 above)
Comment (Tomanhez) [reply features/cart/shopping_cart/removing_cart_item.feature:28]: I had problem, but i fixed it 👍

File: features/cart/shopping_cart/removing_cart_item.feature [hunk 7]
```diff
@@ -7,25 +7,27 @@ Feature: Removing cart item from cart
...
         When I remove product "T-shirt banana" from the cart
         Then my cart should be empty
         And my cart's total should be "$0.00"
 
-    @ui
+    @ui @api
     Scenario: Checking cart's total after removing one item
         Given the store has a product "T-shirt strawberry" priced at "$17.22"
-        And I add this product to the cart
+        And I add "T-shirt banana" product to the cart
```
Comment (lchrusciel) [inline features/cart/shopping_cart/removing_cart_item.feature:30]:
```suggestion
        And I added "T-shirt banana" product to the cart
```

File: features/cart/shopping_cart/removing_cart_item.feature [hunk 8]
```diff
@@ -7,25 +7,27 @@ Feature: Removing cart item from cart
...
+        And I add "T-shirt strawberry" product to the cart
```
Comment (lchrusciel) [inline features/cart/shopping_cart/removing_cart_item.feature:31]:
```suggestion
        And I added "T-shirt strawberry" product to the cart
```

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 9]
```diff
@@ -62,6 +62,8 @@ public function iSeeTheSummaryOfMyCart(string $tokenValue): void
     }
 
     /**
+     * @Given /^I added (product "[^"]+") to the (cart)$/
+     * @Given /^I (?:add|added) ("[^"]+" product) to the (cart)$/
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:66]: The rule of thumb is to not mix `given` and `when` scenarios. Can you extract past tense definitions to `Setup\CartContext` (with the usage of commands instead of API calls)

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Cart/RemoveItemFromCartHandlerSpec.php [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+        $cart->getItemUnitsByVariant($productVariant)->willReturn($orderItemUnits);
+
+        $orderModifier->removeFromOrder($cart, $cartItem)->shouldBeCalled();
+
+        $orderProcessor->process($cart)->shouldBeCalled();
+
+        $this(RemoveItemFromCart::removeFromData('TOKEN', 'PRODUCT_CODE'))->shouldReturn($cart);
+    }
+
+    function it_throws_an_exception_if_product_is_not_found(
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Cart/RemoveItemFromCartHandlerSpec.php:85]:
```suggestion
    function it_throws_an_exception_if_product_was_not_found(
```

File: src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Cart/RemoveItemFromCartHandlerSpec.php [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+            'tokenValue' => 'TOKEN',
+        ])->shouldNotBeCalled();
+
+        $this
+            ->shouldThrow(\InvalidArgumentException::class)
+            ->during('__invoke', [RemoveItemFromCart::removeFromData('TOKEN', 'PRODUCT_CODE')])
+        ;
+    }
+
+    function it_throws_an_exception_if_cart_is_not_found(
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/spec/CommandHandler/Cart/RemoveItemFromCartHandlerSpec.php:102]:
```suggestion
    function it_throws_an_exception_if_cart_was_not_found(
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml [hunk 12]
```diff
@@ -73,6 +73,20 @@
                     <attribute name="groups">cart:add_item</attribute>
                 </attribute>
             </itemOperation>
+
+            <itemOperation name="remove_item">
+                <attribute name="security">is_granted('IS_AUTHENTICATED_ANONYMOUSLY')</attribute>
+                <attribute name="openapi_context">
+                    <attribute name="summary">Remove Item</attribute>
+                </attribute>
+                <attribute name="method">PATCH</attribute>
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml:82]:
```suggestion
                <attribute name="method">DELETE</attribute>
```

File: src/Sylius/Bundle/ApiBundle/DataTransformer/RemoveItemFromCartDataTransformer.php [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\DataTransformer;
+
+use ApiPlatform\Core\DataTransformer\DataTransformerInterface;
+use Sylius\Bundle\ApiBundle\Command\Cart\RemoveItemFromCart;
+use Sylius\Component\Core\Model\OrderInterface;
+use Webmozart\Assert\Assert;
+
+final class RemoveItemFromCartDataTransformer implements DataTransformerInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/DataTransformer/RemoveItemFromCartDataTransformer.php:21]: This is a one-to-one copy-paste of `AddItemToCartInputDataTransformer`, am I right? IF so, maybe we should provide some generic data transformer, which will enrich all token aware commands?

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php (hunk 9 above)
Comment (Tomanhez) [reply src/Sylius/Behat/Context/Api/Shop/CartContext.php:66]: It is not easy because is many differences with using this steps in @ui and @api tests, First problem is properly token transfer between ui steps, because now in @ui context we open product by slug and add it to cart, next steps are continue with this cart, its difficult operate with token from setup context step

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml (hunk 12 above)
Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml:82]: Delete can't has a body. In our case we need add product code to delete, I thing that in this case we cant add DELETE, https://stackoverflow.com/questions/299628/is-an-entity-body-allowed-for-an-http-delete-request

Screenshot 2020-06-30 at 11 22 53

File: src/Sylius/Bundle/ApiBundle/DataTransformer/RemoveItemFromCartDataTransformer.php (hunk 13 above)
Comment (Tomanhez) [reply src/Sylius/Bundle/ApiBundle/DataTransformer/RemoveItemFromCartDataTransformer.php:21]: I will improve it in next PR

File: src/Sylius/Bundle/OrderBundle/Doctrine/ORM/OrderRepository.php [hunk 14]
```diff
@@ -55,6 +55,17 @@ public function findLatest(int $count): array
...
 
+    public function findLatestCart(): ?OrderInterface
+    {
+        return $this->createQueryBuilder('o')
+            ->andWhere('o.state = :state')
+            ->setParameter('state', OrderInterface::STATE_CART)
+            ->addOrderBy('o.checkoutCompletedAt', 'DESC')
+            ->getQuery()
+            ->getOneOrNullResult()
+            ;
```
Comment (AdamKasp) [inline src/Sylius/Bundle/OrderBundle/Doctrine/ORM/OrderRepository.php:66]:
```suggestion
        ;
```

File: src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml (hunk 12 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/ApiBundle/Resources/config/api_resources/Order.xml:82]: Oh, but this IRI structure is wrong. We should rather have `/new-api/orders/{id}/items/{itemId}` with DELETE method on it.

File: src/Sylius/Bundle/ApiBundle/Command/Cart/RemoveItemFromCart.php [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+final class RemoveItemFromCart
+{
+    /** @var string|null */
+    public $tokenValue;
+
+    /**
+     * @var string
+     * @psalm-immutable
+     */
+    public $productCode;
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Command/Cart/RemoveItemFromCart.php:25]: We cannot remove items by `productCode`, we need to provide `orderItemId` and based on this choose a proper item to remove

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 16]
```diff
@@ -77,6 +79,27 @@ public function iAddOfThemToMyCart(int $quantity, ProductInterface $product, str
...
+     * @When /^I remove (product "[^"]+") from the (cart)$/
+     */
+    public function iRemoveProductFromTheCart(ProductInterface $product, string $tokenValue): void
+    {
+        $items = $this->responseChecker->getValue($this->cartsClient->show($tokenValue), 'items');
+
+        foreach ($items as $item) {
+            $pathElements = explode('/', $item['variant']['product']);
+
+            $productVariantCode = $pathElements[array_key_last($pathElements)];
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:92]:
```suggestion
            $productCode = $pathElements[array_key_last($pathElements)];
```
However, I would say, that just making a request for product would be a better solution.

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 17]
```diff
@@ -77,6 +79,27 @@ public function iAddOfThemToMyCart(int $quantity, ProductInterface $product, str
...
+
+            /** @var ProductVariantInterface $productVariant */
+            $productVariant = $product->getVariants()->first();
+
+            if ($productVariant->getCode() === $productVariantCode) {
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:97]:
```suggestion
            if ($product->getCode() === $productCode) {
```

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 18]
```diff
@@ -77,6 +79,27 @@ public function iAddOfThemToMyCart(int $quantity, ProductInterface $product, str
...
+        $items = $this->responseChecker->getValue($this->cartsClient->show($tokenValue), 'items');
+
+        foreach ($items as $item) {
+            $pathElements = explode('/', $item['variant']['product']);
+
+            $productVariantCode = $pathElements[array_key_last($pathElements)];
+
+            /** @var ProductVariantInterface $productVariant */
+            $productVariant = $product->getVariants()->first();
+
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:96]:
```suggestion
```

File: src/Sylius/Bundle/ApiBundle/CommandHandler/Cart/RemoveItemFromCartHandler.php [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+
+        Assert::notNull($orderItem);
+
+        /** @var OrderInterface|null $cart */
+        $cart = $this->orderRepository->findOneBy([
+            'state' => OrderInterface::STATE_CART,
+            'tokenValue' => $removeItemFromCart->tokenValue,
+        ]);
+
+        Assert::notNull($cart);
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/CommandHandler/Cart/RemoveItemFromCartHandler.php:68]: Instead of fetching cart from the repository, we can just check if order related to order item has expected token I suppose
```suggestion
        $cart = $orderItem->getOrder();

        Assert::same($cart->getTokenValue(), $removeItemFromCart->tokenValue);
```

File: src/Sylius/Bundle/OrderBundle/Doctrine/ORM/OrderItemRepository.php [hunk 20]
```diff
@@ -34,4 +34,19 @@ public function findOneByIdAndCartId($id, $cartId): ?OrderItemInterface
...
         ;
     }
+
+    public function findOneByIdAndCartTokenValue($id, $tokenValue): ?OrderItemInterface
+    {
+        return $this->createQueryBuilder('o')
+            ->innerJoin('o.order', 'cart')
+            ->andWhere('cart.state = :state')
+            ->andWhere('o.id = :id')
+            ->andWhere('cart.tokenValue = :tokenValue')
```
Comment (GSadee) [inline src/Sylius/Bundle/OrderBundle/Doctrine/ORM/OrderItemRepository.php:44]:
```suggestion
            ->andWhere('o.id = :id')
            ->andWhere('cart.state = :state')
            ->andWhere('cart.tokenValue = :tokenValue')
```

**Comment by GSadee** [comment]
Thanks, Tomasz! :tada:
